src/main/java/org/openstreetmap/atlas/checks/validation/areas/WaterbodyAndIslandSizeCheck.java

```diff
@@ -0,0 +1,310 @@
+                            object.getOsmIdentifier(), surfaceAreaMeters, this.isletMinimumArea)));
+                }
+                else if (surfaceAreaKilometers > this.isletMaximumArea)
+                {
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(FIVE, object.getOsmIdentifier(),
+                                    surfaceAreaKilometers, this.isletMaximumArea)));
+                }
+            }
+        }
+        // There are two Relations to examine:
+        // 1. Islands included as members of a MultiPolygon (not otherwise tagged as place=island or
+        // place=islet)
+        // 2. Water bodies included as members of a Multipolygon (not otherwise tagged as
+        // natural=water)
+        // Each Relation will have n numbers of instructions, dependent on the amount of members
+        // that fail the size check
+        else if (object instanceof Relation)
+        {
+            final Relation relation = (Relation) object;
+            // Get Relation members that are an Area, have either an inner or outer member,
+            // and have yet to be examined
+            final Set<RelationMember> relationMembers = relation.members().stream()
+                    .filter(member -> member.getEntity() instanceof Area)
+                    .filter(member -> !this.isFlagged(member.getEntity().getIdentifier()))
+                    .filter(this::isInnerOrOuterMember).collect(Collectors.toSet());
+
+            if (!relationMembers.isEmpty())
+            {
+                final List<String> flagInstructions = new ArrayList<>();
+                final Set<RelationMember> invalidRelationMembers = new HashSet<>();
+
+                // Append member and given instruction for Relation members that fail the surface
+                // area size test.
+                this.getInvalidRelationMembers(relationMembers, invalidRelationMembers,
+                        flagInstructions);
+
+                if (!flagInstructions.isEmpty())
+                {
+                    // Create a single flag with each member entity and its given instruction
+                    return Optional.of(flagInstructions).map(instruction -> new CheckFlag(
+                            this.getTaskIdentifier(relation), invalidRelationMembers.stream()
+                                    .map(RelationMember::getEntity).collect(Collectors.toSet()),
+                            instruction));
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Helper function for filtering invalid Relation members. Each Relation member have either an
+     * inner or outer role, and the inner role must not contain the natural=rock tag.
+     *
+     * @param member
+     *            - RelationMember
+     * @return true if member is valid
+     */
+    private boolean isInnerOrOuterMember(final RelationMember member)
```

Comment:
Method is doing more than checking if given member is inner or outer. Member is checked if it is a rock or not. I'd suggest either updating method name or taking rock piece out of the method and moving it into `filter` call.

Reply:
On a second thought, conditions used for filtering on lines 209-211 could be combined in this method.

Reply:
Yeah, good idea